pkg/agent/core/wss.go

```diff
@@ -0,0 +1,87 @@
+package core
+
+// Working set size estimation
+// For more, see: https://www.notion.so/neondatabase/874ef1cc942a4e6592434dbe9e609350
+
+import (
+	"fmt"
+)
+
+type WssEstimatorConfig struct {
+	// MaxAllowedIncreaseFactor is the maximum tolerable increase in slope between windows.
+	// If the slope increases by more than this factor, we will cut off the working set size as the
+	// border between the two windows.
+	MaxAllowedIncreaseFactor float64
+	// InitialOffset is the index of the minimum working set size we must consider.
+	//
+	// In practice, this is taken from the scaling config's LFCMinWaitBeforeDownscaleMinutes, with
+	// the expectation that datapoints are all one minute apart, starting at 1m. So a value of 15m
+	// translates to an InitialOffset of 14 (-1 because indexes start at zero, but the first
+	// datapoint is 1m).
+	InitialOffset int
+	// WindowSize sets the number of datapoints averaged out in a slope calculation.
+	// This value must be >= 2.
+	//
+	// In practice, this value is taken from the scaling config's LFCWindowSizeMinutes, with the
+	// expectation that datapoints are all one minute apart. So, a value of 5 minutes translates to
+	// a WindowSize of 5.
+	WindowSize int
+}
+
+// EstimateTrueWorkingSetSize returns an estimate of the "true" current working set size, given a
+// series of datapoints for the observed working set size over increasing time intervals.
+//
+// In practice, the 'series' is e.g., values of 'neon.lfc_approximate_working_set_size_seconds(d)'
+// for equidistant values of 'd' from 1 minute to 60 minutes.
+func EstimateTrueWorkingSetSize(
+	series []float64,
+	cfg WssEstimatorConfig,
+) float64 {
+	if cfg.WindowSize < 2 {
+		panic(fmt.Errorf("cfg.WindowSize must be >= 2 (got %v)", cfg.WindowSize))
+	}
+
+	w := cfg.WindowSize - 1
+
+	for t := cfg.InitialOffset; t < len(series)-w; t += 1 {
+
+		d0 := max(0.0, series[t]-series[t-w])
```

Comment:
max with 0 is just a precaution, right? From how HLL is designed, this shouldn't be possible, right?

Reply:
Technically, the values in the metric aren't guaranteed to be computed at the same time, so it's in theory possible to get `series[t] < series[t-w]`